fish-rust/src/common.rs

```diff
@@ -3,13 +3,115 @@ use crate::wchar::{wstr, WString};
 use crate::wchar_ext::WExt;
 use crate::wchar_ffi::c_str;
 use crate::wchar_ffi::WCharFromFFI;
+use std::mem::ManuallyDrop;
+use std::ops::{Deref, DerefMut};
 use std::os::fd::AsRawFd;
 use std::{ffi::c_uint, mem};
 
+/// Like [`std::mem::replace()`] but provides a reference to the old value in a callback to obtain
+/// the replacement value. Useful to avoid errors about multiple references (`&mut T` for `old` then
+/// `&T` again in the `new` expression).
+pub fn replace_with<T, F: FnOnce(&T) -> T>(old: &mut T, with: F) -> T {
+    let new = with(&*old);
+    std::mem::replace(old, new)
+}
+
+/// A RAII cleanup object. Unlike in C++ where there is no borrow checker, we can't just provide a
+/// callback that modifies live objects willy-nilly because then there would be two &mut references
+/// to the same object - the original variables we keep around to use and their captured references
+/// held by the closure until its scope expires.
+///
+/// Instead we have a `ScopeGuard` type that takes exclusive ownership of (a mutable reference to)
+/// the object to be managed. In lieu of keeping the original value around, we obtain a regular or
+/// mutable reference to it via ScopeGuard's [`Deref`] and [`DerefMut`] impls.
+///
+/// The `ScopeGuard` is considered to be the exclusively owner of the passed value for the
+/// duration of its lifetime. If you need to use the value again, use `ScopeGuard` to shadow the
+/// value and obtain a reference to it via the `ScopeGuard` itself:
+///
+/// ```rust
+/// use std::io::prelude::*;
+///
+/// let file = std::fs::File::open("/dev/null");
+/// // Create a scope guard to write to the file when the scope expires.
+/// // To be able to still use the file, shadow `file` with the ScopeGuard itself.
+/// let mut file = ScopeGuard::new(file, |file| file.write_all(b"goodbye\n").unwrap());
+/// // Now write to the file normally "through" the capturing ScopeGuard instance.
+/// file.write_all(b"hello\n").unwrap();
+///
+/// // hello will be written first, then goodbye.
+/// ```
+pub struct ScopeGuard<T, F: FnOnce(&mut T)> {
+    captured: ManuallyDrop<T>,
+    on_drop: Option<F>,
+}
+
+impl<T, F: FnOnce(&mut T)> ScopeGuard<T, F> {
+    /// Creates a new `ScopeGuard` wrapping `value`. The `on_drop` callback is executed when the
+    /// ScopeGuard's lifetime expires or when it is manually dropped.
+    pub fn new(value: T, on_drop: F) -> Self {
+        Self {
+            captured: ManuallyDrop::new(value),
+            on_drop: Some(on_drop),
+        }
+    }
+
+    /// Cancel the unwind operation, e.g. do not call the previously passed-in `on_drop` callback
+    /// when the current scope expires.
+    pub fn cancel(guard: &mut Self) {
+        guard.on_drop.take();
+    }
+
+    /// Cancels the unwind operation like [`ScopeGuard::cancel()`] but also returns the captured
+    /// value (consuming the `ScopeGuard` in the process).
+    pub fn rollback(mut guard: Self) -> T {
```

Comment:
I think this leaks `on_drop`, `let _ = self.on_drop;` should be enough to drop it explicitly.

Reply:
Yikes, nice catch!